consensus/src/dag/dag_store.rs

```diff
@@ -78,62 +96,154 @@ impl Dag {
         Ok(())
     }
 
-    pub fn exists(&self, digest: &HashValue) -> bool {
-        self.nodes_by_digest.contains_key(digest)
+    pub fn exists(&self, metadata: &NodeMetadata) -> bool {
+        self.get_node_ref_by_metadata(metadata).is_some()
     }
 
     pub fn all_exists(&self, nodes: &[NodeCertificate]) -> bool {
-        nodes.iter().all(|certificate| {
-            self.nodes_by_digest
-                .contains_key(certificate.metadata().digest())
-        })
+        nodes
+            .iter()
+            .all(|certificate| self.exists(certificate.metadata()))
     }
 
-    pub fn get_node(&self, digest: &HashValue) -> Option<Arc<CertifiedNode>> {
-        self.nodes_by_digest.get(digest).cloned()
+    fn get_node_ref_by_metadata(&self, metadata: &NodeMetadata) -> Option<&NodeStatus> {
+        self.get_node_ref(metadata.round(), metadata.author())
+    }
+
+    fn get_node_ref(&self, round: Round, author: &Author) -> Option<&NodeStatus> {
+        let index = self.author_to_index.get(author)?;
+        let round_ref = self.nodes_by_round.get(&round)?;
+        round_ref[*index].as_ref()
+    }
+
+    fn get_round_iter(&self, round: Round) -> Option<impl Iterator<Item = &NodeStatus>> {
+        self.nodes_by_round
+            .get(&round)
+            .map(|round_ref| round_ref.iter().flatten())
+    }
+
+    pub fn get_node(&self, metadata: &NodeMetadata) -> Option<Arc<CertifiedNode>> {
+        self.get_node_ref_by_metadata(metadata)
+            .map(|node_status| node_status.as_node().clone())
+    }
+
+    pub fn get_node_by_round_author(
+        &self,
+        round: Round,
+        author: &Author,
+    ) -> Option<&Arc<CertifiedNode>> {
+        self.get_node_ref(round, author)
+            .map(|node_status| node_status.as_node())
+    }
+
+    // TODO: I think we can cache votes in the NodeStatus::Unordered
+    pub fn check_votes_for_node(
+        &self,
+        metadata: &NodeMetadata,
+        validator_verifier: &ValidatorVerifier,
+    ) -> bool {
+        self.get_round_iter(metadata.round() + 1)
+            .and_then(|next_round_iter| {
+                let votes = next_round_iter
+                    .filter(|node_status| {
+                        node_status
+                            .as_node()
+                            .parents()
+                            .iter()
+                            .any(|cert| cert.metadata() == metadata)
+                    })
+                    .map(|node_status| node_status.as_node().author());
+                Some(validator_verifier.check_voting_power(votes).is_ok())
+            })
+            .unwrap_or(false)
+    }
+
+    fn reachable_filter(start: HashValue) -> impl FnMut(&Arc<CertifiedNode>) -> bool {
+        let mut reachable = HashSet::from([start]);
+        move |node| {
+            for parent in node.parents() {
+                reachable.insert(*parent.metadata().digest());
+            }
+            reachable.contains(&node.digest())
```

Comment:
I think you should only insert the parents if this node is reachable, otherwise you could add parents without them actually reachable.

Reply:
good find!